nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteMetricsReportingTask.java

```diff
@@ -0,0 +1,212 @@
+    static final AllowableValue RECORD_FORMAT = new AllowableValue("record-format", "Record Format", "Metrics will be formatted"
+            + " using the Record Writer property of this reporting task. See Additional Details in Usage documentation to"
+            + " have the description of the default schema.");
+
+    static final PropertyDescriptor APPLICATION_ID = new PropertyDescriptor.Builder()
+            .name("s2s-metrics-application-id")
+            .displayName("Application ID")
+            .description("The Application ID to be included in the metrics")
+            .required(true)
+            .expressionLanguageSupported(true)
+            .defaultValue("nifi")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    static final PropertyDescriptor HOSTNAME = new PropertyDescriptor.Builder()
+            .name("s2s-metrics-hostname")
+            .displayName("Hostname")
+            .description("The Hostname of this NiFi instance to be included in the metrics")
+            .required(true)
+            .expressionLanguageSupported(true)
+            .defaultValue("${hostname(true)}")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    static final PropertyDescriptor FORMAT = new PropertyDescriptor.Builder()
+            .name("s2s-metrics-format")
+            .displayName("Output format")
+            .description("The output format that will be used for the metrics")
+            .required(true)
+            .allowableValues(AMBARI_FORMAT, RECORD_FORMAT)
+            .defaultValue(AMBARI_FORMAT.getValue())
+            .addValidator(Validator.VALID)
+            .build();
+
+    private final MetricsService metricsService = new MetricsService();
+
+    public SiteToSiteMetricsReportingTask() throws IOException {
+        final InputStream schema = getClass().getClassLoader().getResourceAsStream("schema-metrics.avsc");
+        recordSchema = AvroTypeUtil.createSchema(new Schema.Parser().parse(schema));
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        final List<PropertyDescriptor> properties = new ArrayList<>(super.getSupportedPropertyDescriptors());
+        properties.add(HOSTNAME);
+        properties.add(APPLICATION_ID);
+        properties.add(FORMAT);
+        properties.add(RECORD_WRITER);
+        properties.remove(BATCH_SIZE);
+        return properties;
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(ValidationContext validationContext) {
+        final List<ValidationResult> problems = new ArrayList<>(super.customValidate(validationContext));
+
+        final boolean isWriterSet = validationContext.getProperty(RECORD_WRITER).isSet();
+        if (validationContext.getProperty(FORMAT).getValue().equals(RECORD_FORMAT.getValue()) && !isWriterSet) {
```

Comment:
Can we also check that a Record Writer is NOT set when Ambari Format is selected? This isn't a requirement, just a suggestion. Since it requires two operations to switch back and forth between formats, that's not an ideal user experience, so I'm fine with whichever way you'd like to do it :)

Reply:
I added the check and added unit tests for the customValidate